core/foundation/inc/ROOT/TypeTraits.hxx

```diff
@@ -0,0 +1,203 @@
+   using arg_types_nodecay = typename CallableTraits<decltype(&T::operator())>::arg_types_nodecay;
+   using ret_type = typename CallableTraits<decltype(&T::operator())>::ret_type;
+};
+
+// lambdas and std::function
+template <typename R, typename T, typename... Args>
+struct CallableTraits<R (T::*)(Args...) const> {
+   using arg_types = TypeList<typename std::decay<Args>::type...>;
+   using arg_types_nodecay = TypeList<Args...>;
+   using ret_type = R;
+};
+
+// mutable lambdas and functor classes
+template <typename R, typename T, typename... Args>
+struct CallableTraits<R (T::*)(Args...)> {
+   using arg_types = TypeList<typename std::decay<Args>::type...>;
+   using arg_types_nodecay = TypeList<Args...>;
+   using ret_type = R;
+};
+
+// function pointers
+template <typename R, typename... Args>
+struct CallableTraits<R (*)(Args...)> {
+   using arg_types = TypeList<typename std::decay<Args>::type...>;
+   using arg_types_nodecay = TypeList<Args...>;
+   using ret_type = R;
+};
+
+// free functions
+template <typename R, typename... Args>
+struct CallableTraits<R(Args...)> {
+   using arg_types = TypeList<typename std::decay<Args>::type...>;
+   using arg_types_nodecay = TypeList<Args...>;
+   using ret_type = R;
+};
+
+// Return first of a variadic list of types.
+template <typename T, typename... Rest>
+struct TakeFirst {
+   using type = T;
+};
+
+template <typename... Types>
+using TakeFirst_t = typename TakeFirst<Types...>::type;
+
+// Remove first type from a variadic list of types, return a TypeList containing the rest.
+// e.g. RemoveFirst_t<A,B,C> is TypeList<B,C>
+template <typename T, typename... Rest>
+struct RemoveFirst {
+   using type = TypeList<Rest...>;
+};
+
+template <typename... Args>
+using RemoveFirst_t = typename RemoveFirst<Args...>::type;
+
+/// Return first of possibly many template parameters.
+/// For non-template types, the result is the type itself.
+/// e.g. TakeFirstParameter<U<A,B>> is A
+///      TakeFirstParameter<T> is T
+template <typename T>
+struct TakeFirstParameter {
+   using type = T;
+};
+
+template <template <typename...> class Template, typename T, typename... Rest>
+struct TakeFirstParameter<Template<T, Rest...>> {
+   using type = T;
+};
+
+template <typename T>
+using TakeFirstParameter_t = typename TakeFirstParameter<T>::type;
+
+/// Remove first of possibly many template parameters.
+/// e.g. RemoveFirstParameter_t<U<A,B>> is U<B>
+template <typename>
+struct RemoveFirstParameter {
+};
+
+template <typename T, template <typename...> class U, typename... Rest>
+struct RemoveFirstParameter<U<T, Rest...>> {
+   using type = U<Rest...>;
+};
+
+template <typename T>
+using RemoveFirstParameter_t = typename RemoveFirstParameter<T>::type;
+
+/// Compile-time integer sequence generator
+/// e.g. calling GenStaticSeq<3>::type() instantiates a StaticSeq<0,1,2>
+// TODO deprecate as soon as std::index_sequence is backported
```

Comment:
Actually since it is not meant to be used by users (std::index_sequence is), I would put it in ::ROOT::Internal (and move it back to DataFrame unless there is another 'current' user ...